### PR TITLE
feat: runtime check for schema codegenVersion

### PR DESCRIPTION
**Pull request by dpilch**
#### Description of changes

Check if the `codegenVersion` used for codegen is supported by the current version of DataStore. If it is not supported throw an error. Uses a regex to determine if the semantic version is in valid range.

I did not include configuration for the patch number because anything that would cause the supported `codegenVersion` to increase should be a feature and a minor version bump.

#### Issue #, if available

N/A

#### Description of how you validated changes

Unit tests

#### Checklist

- [x] PR description included
- [ ] `yarn test` passes (new tests pass, but some existing ones fail)
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/datastore/package.json [hunk 1]
```diff
@@ -56,6 +56,7 @@
 		"amazon-cognito-identity-js": "5.2.10",
 		"idb": "5.0.6",
 		"immer": "9.0.6",
+		"semver": "7.3.7",
```
Comment (dpilch) [inline packages/datastore/package.json]: Added a new dep here, but additional size to bundle with partial import would be small `import semverSatisfies from 'semver/functions/satisfies';`. We could roll our own solution but would take a little extra time.

File: packages/datastore/src/datastore/datastore.ts [hunk 2]
```diff
@@ -360,6 +364,20 @@ const checkSchemaInitialized = () => {
 	}
 };
 
+/**
+ * Throws an excpetion if the schema is using a pragma that is not supported.
+ *
+ * @param pragma schema pragma
+ */
+const checkSchemaPragma = (pragma: string) => {
+	const supportedRange = '^3.1.0';
```
Comment (dpilch) [inline packages/datastore/src/datastore/datastore.ts]: Just a random number here now.

File: packages/datastore/src/datastore/datastore.ts [hunk 3]
```diff
@@ -360,6 +364,20 @@ const checkSchemaInitialized = () => {
...
+	if (!semverSatisfies(pragma, supportedRange)) {
+		const message = `Models were generated with an unsuported version of codegen and DataStore will not function as a result. blah blah`;
```
Comment (svidgen) [inline packages/datastore/src/datastore/datastore.ts]:
```suggestion
		const message = `Models were generated with an unsupported version of codegen. Codegen artifacts are from ${pragma || 'an unknown version'}, whereas ${supportedRange} is required. Update to the latest CLI and rerun codegen.`;
```

File: packages/datastore/src/datastore/datastore.ts (hunk 2 above)
Comment (svidgen) [reply packages/datastore/src/datastore/datastore.ts]: Is there a meaningful range we can put here before merging? If we don't know yet, can you add an internal task to update this when the CLI update is released?

**Review by svidgen — CHANGES_REQUESTED** [review]

File: packages/datastore/src/datastore/datastore.ts (hunk 2 above)
Comment (dpilch) [reply packages/datastore/src/datastore/datastore.ts]: `3.1.0` is the latest version so `3.2.0` would be the lowest possible. I'll add a todo and a task.

File: packages/datastore/package.json (hunk 1 above)
Comment (svidgen) [reply packages/datastore/package.json]: Yeah, this is pretty heavy for a version check that we control on both ends. If this blows bundle size, let's start with a more rudimentary check.

For our purposes, we can start by asserting that the "supported version" string is _simple_, in the form of `^N.N.N`, and then that each part of the "version to check" is `>=` the expected part. I don't think we'll need anything more sophisticated than that for quite awhile. And, we can safeguard against bad assumptions by moving the expected version to a constant and writing a test around it to ensure it's a "simple" semver matcher.

**Review by svidgen — APPROVED** [review]

**Pull request merged**